python/tvm/micro/base.py

```diff
@@ -0,0 +1,178 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Base definitions for micro."""
+
+from __future__ import absolute_import
+
+import logging
+import os
+
+import tvm.module
+from tvm.contrib import graph_runtime, util
+from tvm import relay
+
+from .._ffi.function import _init_api
+from .._ffi.libinfo import find_include_path
+from .cross_compile import create_lib
+
+SUPPORTED_DEVICE_TYPES = ["host", "openocd"]
+
+class Session:
+    """MicroTVM Session
+
+    Example
+    --------
+    .. code-block:: python
+
+      c_mod = ...  # some module generated with "c" as the target
+      device_type = "host"
+      with tvm.micro.Session(device_type) as sess:
+          sess.create_micro_mod(c_mod)
+    """
+
+    def __init__(self, device_type, binutil_prefix, port=0):
+        """Stores parameters for initializing a micro device session.
+
+        The session is not initialized until the constructed object is used
+        in a `with` block.
+
+        Parameters
+        ----------
+        device_type : str
+            type of low-level device
+
+        binutil_prefix : str
+            binutil prefix to be used. For example, a prefix of
+            "riscv64-unknown-elf-" means "riscv64-unknown-elf-gcc" is used as
+            the compiler and "riscv64-unknown-elf-ld" is used as the linker,
+            etc.
+
+        port : integer, optional
+            port number of OpenOCD server
+        """
+        if device_type not in SUPPORTED_DEVICE_TYPES:
+            raise RuntimeError("unknown micro device type \"{}\"".format(device_type))
+
+        self.device_type = device_type
+        self.binutil_prefix = binutil_prefix
+        self.port = port
+
+    def micro_build(self, func: relay.Function, params={}):
+        """Create a graph runtime module with a micro device context."""
+        with tvm.build_config(disable_vectorize=True):
+            with relay.build_config(opt_level=3):
+                graph, c_mod, params = relay.build(func, target="c", params=params)
+
+        micro_mod = self.create_micro_mod(c_mod)
+        ctx = tvm.micro_dev(0)
+        mod = graph_runtime.create(graph, micro_mod, ctx)
+        return mod, params
+
+    def create_micro_mod(self, c_mod):
+        """Produces a micro module from a given module.
+
+        Parameters
+        ----------
+        c_mod : tvm.module.Module
+            module with "c" as its target backend
+
+        device_type : str
+            type of low-level device to target
+
+        Return
+        ------
+        micro_mod : tvm.module.Module
+            micro module for the target device
+        """
+        temp_dir = util.tempdir()
+        # Save module source to temp file.
+        lib_src_path = temp_dir.relpath("dev_lib.c")
+        mod_src = c_mod.get_source()
+        with open(lib_src_path, "w") as f:
+            f.write(mod_src)
+        # Compile to object file.
```

Comment:
remove this function, instead, directly support tvm.micro.create_lib(c_mod, "mylib.obj")